module/zfs/bpobj.c

```diff
@@ -242,19 +280,23 @@ bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
 		ASSERT3U(offset, >=, dbuf->db_offset);
 		ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
 
-		bparray = dbuf->db_data;
-		bp = &bparray[blkoff];
+		blkptr_t *bparray = dbuf->db_data;
+		blkptr_t *bp = &bparray[blkoff];
 		err = func(arg, bp, tx);
 		if (err)
 			break;
+
 		if (free) {
-			bpo->bpo_phys->bpo_bytes -=
-			    bp_get_dsize_sync(dmu_objset_spa(bpo->bpo_os), bp);
+			spa_t *spa = dmu_objset_spa(bpo->bpo_os);
+			comp_freed = uncomp_freed = 0;
+			freed = bp_get_dsize_sync(spa, bp);
 			ASSERT3S(bpo->bpo_phys->bpo_bytes, >=, 0);
 			if (bpo->bpo_havecomp) {
-				bpo->bpo_phys->bpo_comp -= BP_GET_PSIZE(bp);
-				bpo->bpo_phys->bpo_uncomp -= BP_GET_UCSIZE(bp);
+				comp_freed = BP_GET_PSIZE(bp);
+				uncomp_freed = BP_GET_UCSIZE(bp);
 			}
+			propagate_space_reduction(bpi, freed,
+			    comp_freed, uncomp_freed);
```

Comment:
it seems really expensive to do this every time through the loop.  Could we do it once before returning?